bson/benchmark_test.go

```diff
@@ -305,3 +322,128 @@ func BenchmarkUnmarshal(b *testing.B) {
 		})
 	}
 }
+
+// The following benchmarks are copied from the Go standard library's
+// encoding/json package.
+
+type codeResponse struct {
+	Tree     *codeNode `json:"tree"`
+	Username string    `json:"username"`
+}
+
+type codeNode struct {
+	Name     string      `json:"name"`
+	Kids     []*codeNode `json:"kids"`
+	CLWeight float64     `json:"cl_weight"`
+	Touches  int         `json:"touches"`
+	MinT     int64       `json:"min_t"`
+	MaxT     int64       `json:"max_t"`
+	MeanT    int64       `json:"mean_t"`
+}
+
+var codeJSON []byte
+var codeBSON []byte
+var codeStruct codeResponse
+
+func codeInit() {
+	f, err := os.Open("testdata/code.json.gz")
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+	gz, err := gzip.NewReader(f)
+	if err != nil {
+		panic(err)
+	}
+	data, err := io.ReadAll(gz)
+	if err != nil {
+		panic(err)
+	}
+
+	codeJSON = data
+
+	if err := json.Unmarshal(codeJSON, &codeStruct); err != nil {
+		panic("json.Unmarshal code.json: " + err.Error())
+	}
+
+	if data, err = json.Marshal(&codeStruct); err != nil {
+		panic("json.Marshal code.json: " + err.Error())
+	}
+
+	if codeBSON, err = Marshal(&codeStruct); err != nil {
+		panic("Marshal code.json: " + err.Error())
+	}
+
+	if !bytes.Equal(data, codeJSON) {
+		println("different lengths", len(data), len(codeJSON))
+		for i := 0; i < len(data) && i < len(codeJSON); i++ {
+			if data[i] != codeJSON[i] {
+				println("re-marshal: changed at byte", i)
+				println("orig: ", string(codeJSON[i-10:i+10]))
+				println("new: ", string(data[i-10:i+10]))
+				break
+			}
+		}
+		panic("re-marshal code.json: different result")
+	}
+}
+
+func BenchmarkCodeUnmarshal(b *testing.B) {
+	b.ReportAllocs()
+	if codeJSON == nil {
+		b.StopTimer()
+		codeInit()
+		b.StartTimer()
+	}
+	b.Run("BSON", func(b *testing.B) {
+		b.RunParallel(func(pb *testing.PB) {
+			for pb.Next() {
+				var r codeResponse
+				if err := Unmarshal(codeBSON, &r); err != nil {
+					b.Fatal("Unmarshal:", err)
+				}
+			}
+		})
+		b.SetBytes(int64(len(codeJSON)))
```

Comment:
I used `codeJSON` since that provides a common reference point between the "BSON" and "JSON" benchmarks (basically it makes the decoding rate in MB/s comparable between the two formats - otherwise they differ due to the differences in encoded size). 

That said, it does not accurately represent the BSON decoding speed - so happy to change it to use `len(codeBSON)`.

Reply:
Ah, interesting, thanks for clarifying the motivation! I am in favor of changing it to `len(codeBSON)` because I think subsequent readers will likely be confused about the intent (as I was) the "sec/op" measurement already gives us a reasonable comparison between JSON/BSON.

Reply:
Good point - changed it to use `codeBSON` instead.